Libplanet/Blockchain/Renderers/DelayedRenderer.cs

```diff
@@ -0,0 +1,293 @@
+        protected ILogger Logger { get; }
+
+        /// <inheritdoc cref="IRenderer{T}.RenderBlock(Block{T}, Block{T})"/>
+        public virtual void RenderBlock(Block<T> oldTip, Block<T> newTip)
+        {
+            _confirmed.TryAdd(oldTip.Hash, 0);
+            DiscoverBlock(newTip);
+        }
+
+        /// <inheritdoc cref="IRenderer{T}.RenderReorg(Block{T}, Block{T}, Block{T})"/>
+        public virtual void RenderReorg(Block<T> oldTip, Block<T> newTip, Block<T> branchpoint)
+        {
+            _confirmed.TryAdd(branchpoint.Hash, 0);
+        }
+
+        /// <summary>
+        /// The callback method which is invoked when the new <see cref="Tip"/> is recognized and
+        /// changed.
+        /// </summary>
+        /// <param name="oldTip">The previously recognized topmost block.</param>
+        /// <param name="newTip">The topmost block recognized this time.</param>
+        /// <returns>A branchpoint between <paramref name="oldTip"/> and <paramref name="newTip"/>
+        /// if the tip change is a reorg.  Otherwise returns <c>null</c>.</returns>
+        protected virtual Block<T>? OnTipChanged(Block<T>? oldTip, Block<T> newTip)
+        {
+            if (oldTip is null)
+            {
+                return null;
+            }
+
+            Block<T>? branchpoint = null;
+            if (!newTip.PreviousHash.Equals(oldTip.Hash))
+            {
+                branchpoint = FindBranchpoint(oldTip, newTip);
+                Renderer.RenderReorg(oldTip, newTip, branchpoint);
+            }
+
+            Renderer.RenderBlock(oldTip, newTip);
+            return branchpoint;
+        }
+
+        private void DiscoverBlock(Block<T> block)
+        {
+            if (_confirmed.ContainsKey(block.Hash))
+            {
+                return;
+            }
+
+            _confirmed.TryAdd(block.Hash, 0);
+
+            HashDigest<SHA256>? prev = block.PreviousHash;
+            Block<T>? newTip = null;
+            do
+            {
+                if (!(prev is HashDigest<SHA256> prevHash &&
+                      Store.GetBlock<T>(prevHash) is Block<T> prevBlock))
+                {
+                    break;
+                }
+
+                uint c = _confirmed.AddOrUpdate(prevHash, k => 1U, (k, v) => v + 1U);
+                Logger.Verbose(
+                    "The block #{BlockIndex} {BlockHash} has {Confirmations} confirmations.",
+                    prevBlock.Index,
+                    prevBlock.Hash,
+                    c
+                );
+
+                if (c >= Confirmations)
+                {
+                    if (newTip is null)
+                    {
+                        newTip = prevBlock;
+                    }
+                    else
+                    {
+#pragma warning disable S1116, SA1106, SA1503
+                        // The below while statement is not "empty"; its condition has an effect.
+                        while (!_confirmed.TryRemove(prevHash, out _));
```

Comment:
Are we need to repeat `.TryRemove()` for deletion completely?

Reply:
Yeah, it's `ConcurrentDictionary<K, V>`'s intended usage.

Reply:
Even so, I think we can't guarantee completeness because it doesn't cover addition after this `while`... is it okay?

Reply:
It's okay if we sure each addition is paired to each call to `DiscoverBlock()`.

Reply:
🆗